### PR TITLE
step 3 for middleware architecture

### PR DESCRIPTION
**Pull request by dai-shi**
step 3 for #513 following #545 and #550

- [x] customizable middleware
- [x] rsc without ssr fixture
- [x] windows support
- [x] cookie middleware and example
- cleanup code 👉 next PR

### CODE REVIEW

**Review by himself65 — COMMENTED** [review]
crashed inside analyzeEntries

**Review by himself65 — CHANGES_REQUESTED** [review]
```diff
Subject: [PATCH] fix
---
Index: e2e/fixtures/rsc-basic/tsconfig.config.json
===================================================================
diff --git a/e2e/fixtures/rsc-basic/tsconfig.config.json b/e2e/fixtures/rsc-basic/tsconfig.config.json
deleted file mode 100644
--- a/e2e/fixtures/rsc-basic/tsconfig.config.json	(revision 29c465342cef32a272d0cb3228d5505d37e4b49f)
+++ /dev/null	(revision 29c465342cef32a272d0cb3228d5505d37e4b49f)
@@ -1,7 +0,0 @@
-{
-  "extends": "./tsconfig.json",
-  "compilerOptions": {
-    "outDir": "./dist/config"
-  },
-  "include": ["waku.config.ts"]
-}
Index: e2e/fixtures/rsc-basic/tsconfig.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/e2e/fixtures/rsc-basic/tsconfig.json b/e2e/fixtures/rsc-basic/tsconfig.json
--- a/e2e/fixtures/rsc-basic/tsconfig.json	(revision 29c465342cef32a272d0cb3228d5505d37e4b49f)
+++ b/e2e/fixtures/rsc-basic/tsconfig.json	(date 1709342708116)
@@ -13,10 +13,5 @@
     "outDir": "./dist",
     "composite": true
   },
-  "include": ["./src"],
-  "references": [
-    {
-      "path": "./tsconfig.config.json"
-    }
-  ]
+  "include": ["./src", "./waku.config.ts"]
 }
```

**Comment by himself65** [comment]
I believe this is a vite related issue, let me check where it comes from

**Comment by himself65** [comment]
OHHH, related: https://github.com/vitejs/vite/issues/15870

**Comment by himself65** [comment]
> GET https://pkg.csb.dev/dai-shi/waku/commit/07b708cd/waku error (ERR_PNPM_FETCH_500). Will retry in 10 seconds. 2 retries left.

Why I get this on my windows

**Comment by himself65** [comment]
```
[Error: [vite:load-fallback] Could not load ..waku.config.ts (imported by src/entries.tsx): ENOENT: no such file or directory, open 'C:\Users\himse\Code\waku\e2e\fixtures\rsc-basic\..waku.config.ts'] {
  errno: -4058,
  code: 'PLUGIN_ERROR',
  syscall: 'open',
  path: 'C:\\Users\\himse\\Code\\waku\\e2e\\fixtures\\rsc-basic\\..waku.config.ts',
  pluginCode: 'ENOENT',
  plugin: 'vite:load-fallback',
  hook: 'load',
  watchFiles: [
    'C:/Users/himse/Code/waku/e2e/fixtures/rsc-basic/src/entries.tsx',
    'C:/Users/himse/Code/waku/e2e/fixtures/rsc-basic/src/components/Box.tsx',
    'C:/Users/himse/Code/waku/e2e/fixtures/rsc-basic/src/components/ClientCounter.tsx',
    'C:/Users/himse/Code/waku/e2e/fixtures/rsc-basic/src/components/App.tsx',
    'C:/Users/himse/Code/waku/node_modules/.pnpm/react-server-dom-webpack@18.3.0-canary-6c3b8dbfe-20240226_react-dom@18.3.0-canary-6c3b8dbfe-2_e2gctke4iglky4qqboxoruwzqq/node_modules/react-server-dom-webpack/server.edge.js',
    'C:/Users/himse/Code/waku/e2e/fixtures/rsc-basic/src/components/ServerPing/actions.ts',
    'C:/Users/himse/Code/waku/e2e/fixtures/rsc-basic/src/components/ServerPing/Counter.tsx',
    'C:/Users/himse/Code/waku/e2e/fixtures/rsc-basic/src/components/ServerPing/index.tsx',
    'C:/Users/himse/Code/waku/packages/waku/dist/client.js',
    'C:\\Users\\himse\\Code\\waku\\packages\\waku\\package.json',
    'C:\\Users\\himse\\Code\\waku\\e2e\\fixtures\\rsc-basic\\package.json'
  ]
}

```

**Comment by himself65** [comment]
```diff
Subject: [PATCH] fix
---
Index: packages/waku/src/lib/plugins/vite-plugin-rsc-entries.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/packages/waku/src/lib/plugins/vite-plugin-rsc-entries.ts b/packages/waku/src/lib/plugins/vite-plugin-rsc-entries.ts
--- a/packages/waku/src/lib/plugins/vite-plugin-rsc-entries.ts	(revision 33954dc8963c4037b41573a9355a673e271bdba0)
+++ b/packages/waku/src/lib/plugins/vite-plugin-rsc-entries.ts	(date 1709414959434)
@@ -1,6 +1,7 @@
 import { existsSync } from 'node:fs';
 import path from 'node:path';
 import type { Plugin } from 'vite';
+import { normalizePath } from 'vite';
 
 const CONFIG_FILE = 'waku.config.ts'; // XXX only ts extension
 
@@ -19,9 +20,8 @@
 }
 `;
   if (existsSync(CONFIG_FILE)) {
-    const file = path.relative(
-      path.dirname(opts.entriesFile),
-      path.resolve(CONFIG_FILE),
+    let file = normalizePath(
+      path.relative(path.dirname(opts.entriesFile), path.resolve(CONFIG_FILE)),
     );
     codeToAdd += `
 export const configPromise = import('${file}').then((m) => m.default);

```

**Comment by himself65** [comment]
On windows, the path will be `..\waku.config.js`, I think maybe Vite will remove `\`.

**Comment by dai-shi** [comment]
> > GET https://pkg.csb.dev/dai-shi/waku/commit/07b708cd/waku error (ERR_PNPM_FETCH_500). Will retry in 10 seconds. 2 retries left.
> 
> Why I get this on my windows

I think it's codesandbox issue.

**Comment by dai-shi** [comment]
https://github.com/dai-shi/waku/pull/557#issuecomment-1974914324 fixes it. pretty interesting.
Thanks a lot @himself65 !

File: packages/waku/src/server.ts [hunk 1]
```diff
@@ -57,6 +58,7 @@ export type EntriesDev = {
 };
 
 export type EntriesPrd = EntriesDev & {
+  configPromise: Promise<Config>;
```
Comment (dai-shi) [inline packages/waku/src/server.ts:61]: Should this be `loadConfig` ?

File: packages/waku/src/config.ts [hunk 2]
```diff
@@ -79,6 +83,16 @@ export interface Config {
...
+  /**
+   * Middleware to use
+   * Defaults to:
+   * (cmd: 'dev' | 'start') => [
+   *   ...(cmd === 'dev' ? [import('waku/middleware/dev-server')] : []),
+   *   import('waku/middleware/ssr'),
+   *   import('waku/middleware/rsc'),
+   * ]
+   */
+  middleware?: (cmd: 'dev' | 'start') => Promise<{ default: Middleware }>[];
```
Comment (dai-shi) [inline packages/waku/src/config.ts:95]: Is it better to do `loadMiddleware: (cmd) => Promise<Middleware[]>`?

File: packages/website/package.json [hunk 3]
```diff
@@ -18,7 +18,7 @@
     "react": "18.3.0-canary-6c3b8dbfe-20240226",
     "react-dom": "18.3.0-canary-6c3b8dbfe-20240226",
     "react-server-dom-webpack": "18.3.0-canary-6c3b8dbfe-20240226",
-    "waku": "https://pkg.csb.dev/dai-shi/waku/commit/07b708cd/waku"
+    "waku": "workspace:*"
```
Comment (dai-shi) [inline packages/website/package.json:21]: We'll go back to csb version later, or maybe "file" one.

**Pull request merged**